execute/aggregate.go

```diff
@@ -1,33 +1,159 @@
+		n.Finish(m.SrcDatasetID(), m.Error())
+		return nil
+	case ProcessChunkMsg:
+		return n.processChunk(m.TableChunk())
+	case FlushKeyMsg:
+		return n.flushKey(m.Key())
+	case ProcessMsg:
+		return n.Process(m.SrcDatasetID(), m.Table())
+	}
+	return nil
+}
 
-	config AggregateConfig
+// Process is implemented to remain compatible with legacy upstreams.
+// It converts the incoming stream into a set of appropriate messages.
+func (n *aggregateTransformation) Process(id DatasetID, tbl flux.Table) error {
+	if tbl.Empty() {
+		// Since the table is empty, it won't produce any column readers.
+		// Create an empty buffer which can be processed instead
+		// to force the creation of state.
+		buffer := arrow.EmptyBuffer(tbl.Key(), tbl.Cols())
+		chunk := table.ChunkFromBuffer(buffer)
+		if err := n.processChunk(chunk); err != nil {
+			return err
+		}
+	} else {
+		if err := tbl.Do(func(cr flux.ColReader) error {
+			chunk := table.ChunkFromReader(cr)
+			return n.processChunk(chunk)
+		}); err != nil {
+			return err
+		}
+	}
+	return n.flushKey(tbl.Key())
+}
+
+func (n *aggregateTransformation) processChunk(chunk table.Chunk) error {
+	state, _ := n.d.Lookup(chunk.Key())
+	if ns, ok, err := n.t.Aggregate(chunk, state, n.d.mem); err != nil {
+		return err
+	} else if ok {
+		n.d.Set(chunk.Key(), ns)
+	}
+	return nil
 }
 
-type AggregateConfig struct {
+func (n *aggregateTransformation) flushKey(key flux.GroupKey) error {
+	// Remove the state for this key from the dataset.
+	// If we find state associated with the key, compute the table.
+	if state, ok := n.d.Delete(key); ok {
+		if err := n.t.Compute(key, state, n.d, n.d.mem); err != nil {
+			return err
+		}
+		return n.d.FlushKey(key)
+	}
+	return nil
+}
+
+// Finish is implemented to remain compatible with legacy upstreams.
+func (n *aggregateTransformation) Finish(id DatasetID, err error) {
+	if err == nil {
+		err = n.d.Range(func(key flux.GroupKey, value interface{}) error {
+			if err := n.t.Compute(key, value, n.d, n.d.mem); err != nil {
+				return err
+			}
+			return n.d.FlushKey(key)
+		})
+	}
+	n.d.Finish(err)
+}
+
+func (n *aggregateTransformation) OperationType() string {
+	return OperationType(n.t)
+}
+func (n *aggregateTransformation) RetractTable(id DatasetID, key flux.GroupKey) error {
+	return nil
+}
+func (n *aggregateTransformation) UpdateWatermark(id DatasetID, t Time) error {
+	return nil
+}
+func (n *aggregateTransformation) UpdateProcessingTime(id DatasetID, t Time) error {
+	return nil
+}
+
+type simpleAggregateTransformation struct {
+	agg    SimpleAggregate
+	config SimpleAggregateConfig
+}
+
+type SimpleAggregateConfig struct {
 	plan.DefaultCost
 	Columns []string `json:"columns"`
 }
 
-var DefaultAggregateConfig = AggregateConfig{
+var DefaultSimpleAggregateConfig = SimpleAggregateConfig{
 	Columns: []string{DefaultValueColLabel},
 }
 
-func (c AggregateConfig) Copy() AggregateConfig {
+func (c SimpleAggregateConfig) Copy() SimpleAggregateConfig {
 	nc := c
```

Comment:
newConfig and config will make it more readable than nc, c

Reply:
This is previously existing code that wasn't modified by this change. I'd prefer not to touch the code if that's ok.